pkg/core/diamond_commit.go

```diff
@@ -0,0 +1,326 @@
+package core
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	iradix "github.com/hashicorp/go-immutable-radix"
+	"github.com/oneconcern/datamon/pkg/core/status"
+	"github.com/oneconcern/datamon/pkg/errors"
+	"github.com/oneconcern/datamon/pkg/model"
+	"github.com/oneconcern/datamon/pkg/storage"
+	"github.com/segmentio/ksuid"
+	"go.uber.org/zap"
+)
+
+// Commit a diamond
+func (d *Diamond) Commit(opts ...Option) error {
+	return d.implCommit(opts...)
+}
+
+// Cancel a diamond
+func (d *Diamond) Cancel() error {
+	// TODO(fred)
+	return status.ErrNotImplemented.WrapMessage("diamond cancel")
+}
+
+// WithState returns the diamond with its state modified and transition timestamp updated
+func (d *Diamond) WithState(state model.DiamondState) *Diamond {
+	d.DiamondDescriptor.State = state
+	d.DiamondDescriptor.StateChangeTime = time.Now()
+
+	switch state {
+	case model.DiamondDone, model.DiamondCancelled:
+		d.DiamondDescriptor.EndTime = d.DiamondDescriptor.StateChangeTime
+	}
+	return d
+}
+
+// collectSplits which are done, sorted by completion time
+func (d *Diamond) collectSplits(opts ...Option) (model.SplitDescriptors, error) {
+	splits := make(model.SplitDescriptors, 0, typicalSplitsNum)
+
+	err := ListSplitsApply(
+		d.RepoID,
+		d.DiamondDescriptor.DiamondID,
+		d.contextStores,
+		func(sd model.SplitDescriptor) error {
+			if sd.State == model.SplitDone {
+				splits = append(splits, sd)
+			}
+			return nil
+		},
+		opts...,
+	)
+	if err != nil {
+		return nil, errors.New("couldn't collect splits").Wrap(err)
+	}
+	return splits, nil
+}
+
+func (d *Diamond) checkBundleID() error {
+	if d.BundleID == "" {
+		return nil
+	}
+
+	// case of bundleID preservation (e.g. data migration use case)
+	id, err := ksuid.Parse(d.BundleID)
+	if err != nil {
+		return status.ErrInvalidKsuid.Wrap(err)
+	}
+	d.setBundleID(id.String())
```

Comment:
specifically, this line could be removed, except for the data migration?

is it worth feature-flagging this use case?

Reply:
added feature guard